// zhlint@0.7.3/lib/rules/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultConfig = void 0;
const space_trim_1 = __importDefault(require("./space-trim"));
const punctuation_width_1 = __importDefault(require("./punctuation-width"));
const punctuation_unification_1 = __importDefault(require("./punctuation-unification"));
const case_abbrs_1 = __importDefault(require("./case-abbrs"));
const space_hyper_mark_1 = __importDefault(require("./space-hyper-mark"));
const space_code_1 = __importDefault(require("./space-code"));
const space_letter_1 = __importDefault(require("./space-letter"));
const space_punctuation_1 = __importDefault(require("./space-punctuation"));
const space_quotation_1 = __importDefault(require("./space-quotation"));
const space_bracket_1 = __importDefault(require("./space-bracket"));
const case_linebreak_1 = __importDefault(require("./case-linebreak"));
const case_zh_units_1 = __importDefault(require("./case-zh-units"));
const case_html_entity_1 = __importDefault(require("./case-html-entity"));
const generateHandlers = (options) => {
    return [
        (0, space_trim_1.default)(options),
        (0, punctuation_width_1.default)(options),
        (0, punctuation_unification_1.default)(options),
        (0, case_abbrs_1.default)(options),
        (0, space_hyper_mark_1.default)(options),
        (0, space_code_1.default)(options),
        (0, space_letter_1.default)(options),
        (0, space_punctuation_1.default)(options),
        (0, space_quotation_1.default)(options),
        (0, space_bracket_1.default)(options),
        (0, case_linebreak_1.default)(options),
        (0, case_zh_units_1.default)(options),
        (0, case_html_entity_1.default)(options)
    ];
};
exports.defaultConfig = {
    noSinglePair: true,
    halfwidthPunctuation: `()[]{}`,
    fullwidthPunctuation: `，。：；？！“”‘’`,
    adjustedFullwidthPunctuation: `“”‘’`,
    unifiedPunctuation: 'simplified',
    spaceBetweenHalfwidthContent: true,
    noSpaceBetweenFullwidthContent: true,
    spaceBetweenMixedwidthContent: true,
    noSpaceBeforePauseOrStop: true,
    spaceAfterHalfwidthPauseOrStop: true,
    noSpaceAfterFullwidthPauseOrStop: true,
    spaceOutsideHalfwidthQuotation: true,
    noSpaceOutsideFullwidthQuotation: true,
    noSpaceInsideQuotation: true,
    spaceOutsideHalfwidthBracket: true,
    noSpaceOutsideFullwidthBracket: true,
    noSpaceInsideBracket: true,
    spaceOutsideCode: true,
    noSpaceInsideHyperMark: true,
    trimSpace: true,
    skipZhUnits: `年月日天号时分秒`,
    skipAbbrs: [
        'Mr.',
        'Mrs.',
        'Dr.',
        'Jr.',
        'Sr.',
        'vs.',
        'etc.',
        'i.e.',
        'e.g.',
        'a.k.a.'
    ]
};
exports.default = generateHandlers;
